### ISSUE TITLE
feature: Support tiered storage for NutsDB

### ISSUE THREAD

**Issue by azhsmesos** [nutsdb/nutsdb#260]
When the disk is not enough, we can unload the data to external cheap storage, such as GFS, S3, AWS, etc., to achieve longer storage time. At the same time, we can reduce the memory pressure, and realize clear operation, calculating the time range through [meta file](https://github.com/nutsdb/nutsdb/issues/225), such as we only save the data for a week, etc.?
- task1: implement tiered storage interface.
- task2: implement clear data file.

**Comment by xujiajun** [comment]
I personally feel that this feature is suitable for the scenario of storing some cold data that is not frequently used, so that the cold data can be stored externally and cheaply.

**Comment by bokket** [comment]
[OpenDAL](https://github.com/apache/incubator-opendal) is a data access layer that allows users to easily and efficiently retrieve data from various storage services in a unified way.It has supported many [services](https://github.com/apache/incubator-opendal/issues/5).If it's integrated, there's some direction for the future of this.

**Comment by bigboss2063** [comment]
> [OpenDAL](https://github.com/apache/incubator-opendal) is a data access layer that allows users to easily and efficiently retrieve data from various storage services in a unified way.It has supported many [services](https://github.com/apache/incubator-opendal/issues/5).If it's integrated, there's some direction for the future of this.

This seems like a good idea. Is it convenient to use opendal in Golang currently? Can you elaborate on your thoughts?

**Comment by bokket** [comment]
> > [OpenDAL](https://github.com/apache/incubator-opendal) is a data access layer that allows users to easily and efficiently retrieve data from various storage services in a unified way.It has supported many [services](https://github.com/apache/incubator-opendal/issues/5).If it's integrated, there's some direction for the future of this.
> 
> This seems like a good idea. Is it convenient to use opendal in Golang currently? Can you elaborate on your thoughts?

No,it's [go binding](https://github.com/apache/incubator-opendal/issues/3205) is still in the preliminary stages.It requires `dlopen`, which binds to the `c API library`, and doesn't [release a go package](https://github.com/apache/incubator-opendal/issues/1472).
1. DAL is better suited for s3 hot and cold processing, it is object based but can be extended to rocksdb, postgresql databases.
2. Another direction is to implement on ssdb like pika.
3. This could be put on hold a little bit, I just think DAL might be the way to go to save on storage costs.

**Comment by bigboss2063** [comment]
If [go-storage](https://github.com/beyondstorage/go-storage) is currently stable, I think we can consider using it 😁. Or if you are familiar with it, We can fork it into our organization and ocntinue maintenance.

We can conduct more research before making a decision. @bokket
